lib/msal-core/src/UserAgentApplication.ts

```diff
@@ -388,201 +378,529 @@ export class UserAgentApplication {
+        if (this._tokenReceivedCallback) {
+            this._tokenReceivedCallback(ErrorDescription.inputScopesError, null, ErrorCodes.inputScopesError, Constants.accessToken, this.getUserState(this._cacheStorage.getItem(Constants.stateLogin, this.storeAuthStateInCookie)));
+        return;
+      }
+    }
 
-              var loginStartPage = this._cacheStorage.getItem(Constants.angularLoginRequest);
-              if (!loginStartPage || loginStartPage === "") {
-                  loginStartPage = window.location.href;
-              }
-              else {
-                  this._cacheStorage.setItem(Constants.angularLoginRequest, "");
-              }
+    if (scopes) {
+      scopes = this.filterScopes(scopes);
+    }
+
+    // Get the user object if a session exists
+    const userObject = user ? user : this.getUser();
+
+    // If already in progress, do not proceed
+    // TODO: Should we throw or return an error here?
+    if (this._acquireTokenInProgress) {
+      return;
+    }
+    
+    // If no session exists, prompt the user to login.
+    const scope = scopes.join(" ").toLowerCase();
+    if (!userObject && !(extraQueryParameters && (extraQueryParameters.indexOf(Constants.login_hint) !== -1 ))) {
+    // TODO: This should be replaced with error callback
+    // TODO: Is this always accessToken?
+      if (this._tokenReceivedCallback) {
+        this._logger.info("User login is required");
+        this._tokenReceivedCallback(ErrorDescription.userLoginError, null, ErrorCodes.userLoginError, Constants.accessToken, this.getUserState(this._cacheStorage.getItem(Constants.stateLogin, this.storeAuthStateInCookie)));
+        return;
+      }
+    }
+
+    // Track the acquireToken progress
+    this._acquireTokenInProgress = true;
+
+    let authenticationRequest: AuthenticationRequestParameters;
+    let acquireTokenAuthority = authority ? AuthorityFactory.CreateInstance(authority, this.validateAuthority) : this.authorityInstance;
+
+    // TODO: Set response type here
+    acquireTokenAuthority.ResolveEndpointsAsync().then(() => {
+      // On Fulfillment
+      if (Utils.compareObjects(userObject, this.getUser())) {
+        if (scopes.indexOf(this.clientId) > -1) {
+          authenticationRequest = new AuthenticationRequestParameters(acquireTokenAuthority, this.clientId, scopes, ResponseTypes.id_token, this.getRedirectUri(), this._state);
+        } else {
+          authenticationRequest = new AuthenticationRequestParameters(acquireTokenAuthority, this.clientId, scopes, ResponseTypes.token, this.getRedirectUri(), this._state);
+        }
+      } else {
+        authenticationRequest = new AuthenticationRequestParameters(acquireTokenAuthority, this.clientId, scopes, ResponseTypes.id_token_token, this.getRedirectUri(), this._state);
+      }
+
+      // Cache nonce
+      this._cacheStorage.setItem(Constants.nonceIdToken, authenticationRequest.nonce, this.storeAuthStateInCookie);
+
+      // Cache acquireTokenUserKey
```

Comment:
This could all be broken out to a helper method

Reply:
It is. All consistent patterns across code, trying to box them into helpers. Next 2-3 PRs on refactoring would consist those changes.